torch/nn/modules/module.py

```diff
@@ -1278,8 +1278,34 @@ def _save_to_state_dict(self, destination, prefix, keep_vars):
         if getattr(self.__class__, "get_extra_state", Module.get_extra_state) is not Module.get_extra_state:
             destination[extra_state_key] = self.get_extra_state()
 
-    # The user can pass an optional arbitrary mappable object to `state_dict`, in which case `state_dict` returns
-    # back that same object. But if they pass nothing, an `OrederedDict` is created and returned.
+    def _state_dict_impl(self, destination, prefix, keep_vars):
+        r"""Holds the actual implementation of
+        :meth:`~torch.nn.Module.state_dict`, with recursive calls for
+        descendants of this module.
+
+        In rare cases, users can call this directly to provide a custom
+        :attr:`destination`.
+
+        Args:
+            destination (dict): a dict where state will be stored
+            prefix (str): the prefix for parameters and buffers used in this
+                module
+            keep_vars (bool): whether NOT to return buffers detached from
+                autograd
+        """
+        destination._metadata[prefix[:-1]] = local_metadata = dict(version=self._version)
+        self._save_to_state_dict(destination, prefix, keep_vars)
+        for name, module in self._modules.items():
+            if module is not None:
+                module.state_dict(destination, prefix + name + '.', keep_vars=keep_vars)
```

Comment:
@lkct , @albanD , the way this is implemented here means that whenever a module has a sub-module, calling `state_dict()` (a perfectly fine and non-deprecated thing to do) will raise the deprecation warning. 

Minimal repro: 
```
from torch.nn import Module

m = Module()
m.state_dict()  # this is fine

m.submodule = Module()
m.state_dict()  # this raises the DeprecationWarning (falsely)
```

This should result in this DeprecationWarning being raised left and right in everyone's code who has nested modules (which is pretty much everyone). This is already happening to us. 

cc @dme65, @saitcakmak